### PR TITLE
Port PR 3689 into release 1.5

### PR DESCRIPTION
**Pull request by vladima**

### CODE REVIEW

File: src/services/shims.ts [hunk 1]
```diff
@@ -242,6 +245,12 @@ module ts {
             return createTextChangeRange(
                 createTextSpan(decoded.span.start, decoded.span.length), decoded.newLength);
         }
+
+        public dispose(): void {
+            if ("dispose" in this.scriptSnapshotShim) {
```
Comment (DanielRosenwasser) [inline src/services/shims.ts]: Why do this instead of the property check?

Comment (JsonFreeman) [inline src/services/shims.ts]: Can you do a property check on a COM object?

File: src/services/services.ts [hunk 2]
```diff
@@ -1873,6 +1876,16 @@ module ts {
...
                     // drop it so it can be lazily recreated later
                     newSourceFile.nameTable = undefined;
+
+                    // dispose all resources held by old script snapshot
+                    if (sourceFile !== newSourceFile && sourceFile.scriptSnapshot) {
+                        if (sourceFile.scriptSnapshot.dispose) {
+                            sourceFile.scriptSnapshot.dispose();
+                        }
+
+                        sourceFile.scriptSnapshot = undefined;
```
Comment (JsonFreeman) [inline src/services/services.ts]: We were never even setting scriptSnapshot to undefined?

File: src/services/shims.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/services/shims.ts]: If this is a problem with COM objects, let's leave a comment explaining it.

Comment (JsonFreeman) [inline src/services/shims.ts]: You mean a COMment

File: src/services/services.ts (hunk 2 above)
Comment (vladima) [inline src/services/services.ts]: Yes

File: src/services/shims.ts (hunk 1 above)
Comment (vladima) [inline src/services/shims.ts]: property check here will effectively become method call with no arguments

**Comment by JsonFreeman** [comment]
:+1:

**Comment by mhegazy** [comment]
we need to merge this into master.

**Comment by JsonFreeman** [comment]
I can merge it.

**Comment by JsonFreeman** [comment]
Done

**Pull request merged**